### DISCUSSION TITLE
Real-Time-Carbon-Footprint-Proposal

### DISCUSSION THREAD

**Discussion by seanmcilroy29** [Green-Software-Foundation/sci#354] [General]
From Adrian Cockcroft: 
I have been working on something I'd like to share and discuss at the meeting this week. I've got a proposal for a Real-Time energy and carbon API standard that would hopefully provide the information needed to do SCI and other work in the cloud. I'm presenting at QCon London on March 29th and will share it publicly there. I'm hoping to get the cloud providers to at least consider it as a future direction to work towards, and it would appear as something like CloudWatch metrics from AWS or as a Prometheus Exporter. Summary of the data that is currently available, and the data we'd like to be available from the three leading cloud providers, is in the attached doc. I've shared it with MSFT, AWS and GCP folk in the last few days. This is in no way intended to be a finished standard proposal, and there is no buy-in from anyone yet, it's an initial SWAG at what I think is needed, and which could be an end point to work towards over the next few years.

**Real-Time-Carbon-Footprint-Proposal** - https://drive.google.com/file/d/1vGiZM6R8ZkXxR6M7ELFqI0cAJM9JKP_f/view?usp=share_link

**Comment by adrianco** [comment]
Thanks for the discussion today. I need to do a longer writeup of the idea than is on the slides, but I think my starting point would be a Prometheus Exporter configured to generate a proposed schema, with simulated data sources to start with, and a configurable model that included whatever data is available. Someone mentioned a possibly similar energy exporter might already exist.

**Reply by DavidElliman** [reply]
Adrian, what would be the units of measurement from the point of view of execution? Do you propose a trailing attribution of energy spent and align it with production or sample it in real-time? I had been thinking about PpW and EPI (Intel's) samples and extrapolation as a model.

**Reply by mrchrisadams** [reply]
hi @adrianco, 

Thanks for creating the deck and bringing this up - it's realy helpful!

> Someone mentioned a possibly similar energy exporter might already exist.

There are three possible projects I can think of that might do this:

**[Scaphandre](https://github.com/hubblo-org/scaphandre)** if you have full access to hardware, Scaphandre goes some of the way towards creating the energy exporter as you described. It provides process-level energy readings, and container level readings, and is written in Rust. 
**[Joular](https://github.com/joular/powerjoular/)** - is a similar tool written in Ada. It was recently announced with a recent paper to go with it (see the github repo for the link to the paper). I think it provides similar process level readings as Scaphandre, and is similarly lightweight, but I'm not sure if it exports readings in a prometheus exporter format.
**[Kepler](https://github.com/sustainable-computing-io/kepler)** - Kepler also exports power usage figures, and relies on e-BPF to do it quickly too. It's focussed on Kubernetes.

There are two OSS projects I know that might deal with the carbon figures too:

**[Tracarbon](https://github.com/fvaleye/tracarbon)**  - I think this uses scaphandre under the hood to expose the same metrics, and combines it with numbers from a series of co2 intensity providers
**[Grid intensity Go](https://github.com/thegreenwebfoundation/grid-intensity-go/)** - I worked on this while back with some other folks, assuming that projects like Scaphandre were already dealing with the energy per container/process problem - we thought that you could use something like Grafana/Prometheus to create some of the desired metrics by combining the numbers coming from Scaphandre, with CO2 intensity readings for each location if you had an exporter for carbon intensity that exposed the figures for each location. This library bundles some open data for carbon intensity (national, annual, avg and marginal figures), and also consumes realtime data from a few sources like Electricity Map(s), exposing those via an prometheus exporter.

Those would provide the location based figures, but I don't know of any places that provide market based _and_ location based in metrics at the same time, and at any resolution greater than annual.

I think you could plausibly get these market based numbers from EnergyTag compliant carbon certificate issuers.

The EnergyTag API spec has a concept of certificate 'bundles' - one potential use-case is issuing a query to check for the existence of a bundle of hourly certificates for a given period of time for clean generation from a specific device (like a wind turbine or solar array), that you could match to usage from the grid, as supporting evidence of making a market based scope 2 adjustment.

The link below points to the draft API spec to query for a 'bundle' of hourly certificates, that you might match to usage at a given hour, for a given device. I've linked to the starting of open API spec and the specific bundle endpoint:

https://energytag.org/fea/
https://energytag.org/fea/#tag/Certificates/operation/query_certificate_bundles_certificates_certificates_query_get

The energy tag papers go into quite a of detail about how you might convert annual environmental attribute certificates used for market based scope 2 reporting, to the more granular certificates - I think this largely means you don't need to wait for the issuer of certificates in a given country to adopt the new standards if there are compliant providers operating in the same market.

If you had location based intensity figures already from some current providers, AND those more granular certificates, I think you'd get a decent part of the way to having the metrics you outlined on page 7.

**What about Carbon Aware SDK?**

I might be wrong on this, but I think the [carbon aware SDK](https://github.com/Green-Software-Foundation/carbon-aware-sdk/) is more focussed on _consuming_ these numbers from various cloud providers to help make scheduling decisions, rather than being reference implementation to ask folks to  _produce_ these numbers for consumption in a TSDB, and creating your own reporting. 

That said, I'm not too familiar with it - I'd defer to the folks working on it, as it may well support this use case.

I'm more involved with the policy WG work than the SCI, but I'd be happy to chat more if any of this is of interest / use for your proposal.

**Reply by adrianco** [reply]
Thanks Chris, this is very useful. I agree with your points - this proposal is focused on defining the underlying information that we want cloud providers to generate, which would be an input into most of the other GSF projects. By creating an implementation of a standard interface, we can start building things on top of it, while we wait for better data to arrive from more providers, and to show the value of providing more data.

I've written up a more detailed proposal here as a draft blog post.
https://medium.com/@adrianco/proposal-for-a-realtime-carbon-footprint-standard-60b71c269948

I'm flying to the UK on Weds 22nd, and will be there for a week, ending up at QCon London on the 29th.

**Reply by jawache** [reply]
I also know that @Sealjay was proposing some work to formalize at least internally at the GSF a carbon data interchange format and there might be some overlap between a related project at LF energy. Chris do you have any links to share on this thread that might aid discussion?

**Reply by Sealjay** [reply]
@jawache Specifically for a Carbon Data Interchange Format, and defining the underlying information, @KnowledgeRatio has been working on this project:
https://github.com/Green-Software-Foundation/opensource-wg/issues/60

We agreed in the OSS WG to define the metrics and ontology we need, and use @navveenb's demo next week to compare the two and see if anything is missing.

During some desk research, it looked like LF Energy have this project https://github.com/carbon-data-specification
So we're likely going to see if we can extend that, if it has gaps.

FYI @dtoakley @dtoakley-tw

**Reply by adrianco** [reply]
I took a closer look at Kepler, and think it looks like a good basis to prototype the interface/schema standard. The cloud providers need to supply some underlying metrics to make this work. AWS blocks detailed power measurement for instances to prevent security exploits that use power as a side channel to break encryption, but I think they could provide one minute granularity power data that wouldn't expose the security risk. I'm not sure at this point if Azure and GCP block those interfaces as well.

**Reply by Sealjay** [reply]
For the data format, I strongly think we should extend existing projects like the lf energy specification; otherwise again we will end up with competing specifications. That’s where the convention work landed.
There are many tools monitoring carbon already, see this issue again:
https://github.com/Green-Software-Foundation/opensource-wg/issues/60

We could ask teams from those tools to work with you on this.
@Henry-WattTime @jawache @adrianco

**Reply by Sealjay** [reply]
Bumping this again, prior to the SC call. I don't think we need the https://github.com/Green-Software-Foundation/opensource-wg/issues/60 project if we also have a real time carbon footprint project.

**Reply by KnowledgeRatio** [reply]
Reviewed the material and agree that pursuing both would be a near complete overlap of work and standards. I'm happy to defer to @adrianco's initiative given the analysis that is already present. That said, I would stress the need to look at other open specifications that offer this, and particularly LFEnergy:

https://github.com/carbon-data-specification
http://carbondataspec.org/

**Reply by Sealjay** [reply]
It may be one of the shortest GSF projects in our (also short) history- but have tabled a discussion in today's OSS WG to close the convention project:
https://github.com/Green-Software-Foundation/opensource-wg/issues/83

@KnowledgeRatio if you can't make it, could you add a comment please?

**Reply by KnowledgeRatio** [reply]
I would support closing the CDIF project because I feel that there is more progress and also a larger outcome (real-time) that is to be achieved by the other project. I offer to support/contribute to the RTCF project when that is off the ground. What I would like to see from RTCF is to go beyond the CSP services and take a bigger picture view of carbon data/impact.

I will also attach the requirements I was able to draft and prioritise which could be useful for posterity and the superseding project.

**Comment by plnoel** [comment]
Thank you for formalizing and structuring this proposal. Totally supporting this initiative, we are definitely suffering from the lack of insight on the energy use of the various services we use with the cloud providers to optimize our footprint.

**Comment by jawache** [comment]
I think one interesting characteristic of this standard is that it's a standard for **monitoring** carbon emissions, i think that should be something that clearly surfaces from this proposal, the SCI can be used for monitoring but not necessary, this is primarily focused on monitoring.

So reducing the distance between someone making a change to a software system and then seeing the impact of that change, it's could have the same impact on the sustainability industry as moving from say a yearly release waterfall model to daily releases (CI/CD) did to the software industry. Instead of a figure every year and then trying to figure out what changes drove that number, if the figure comes every minute you'll be able to correlate changes to the system to changes in carbon emissions, which is a much stronger driver of behavioral change than a yearly regulatory report.

**Comment by Henry-WattTime** [comment]
@Oleg-Zhymolokhov @srini1978 @navveenb please share this with the Carbon QL team for their review and potential collaboration.

**Comment by adrianco** [comment]
Update: I took off on some vacation time so haven't made much progress. My plan is to contact the cloud providers and get their reactions. Also Kubecon was last week, and now that's over I will start a discussion with the Kepler project and other CNCF working groups that are relevant. I had a meeting with AWS scheduled yesterday but it was cancelled as they were dealing with the layoffs that dropped (my contact is still there).

**Comment by adrianco** [comment]
There's some relevant work on the Kepler discussion thread I started. Vincent Caldera from Red Hat has built a real time carbon monitoring dashboard for an OS-Climate Kubernetes cluster (running the OpenShift based OSC Data Commons), and shows the results here: https://www.youtube.com/watch?v=dd-6rqp_qlA

**Comment by adrianco** [comment]
I've made some contacts with AWS and Microsoft people so far, and am not in a hurry, it's going to take time to get everyone aligned around something useful. My next step is to flesh out my proposal in a bit more detail and to convene a discussion at GSF about the right place/process to take this forward. I'm also thinking about writing an Amazon PRFAQ form of the proposal - which would iterate as we get more feedback, but help with engaging the relevant AWS people.

**Comment by adrianco** [comment]
I've written this up as a PR-FAQ - have passed it by some people at AWS, Azure and GCP, but not had substantive feedback yet. Positive feedback from Harness.io and CloudZero. https://github.com/Green-Software-Foundation/pr-faqs/pull/10
